File: examples/project-template/src/main/java/be/kuleuven/csa/controller/BeheerBoerderijenController.java
package be.kuleuven.csa.controller;

import javafx.beans.property.ReadOnlyObjectWrapper;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.stage.Stage;

import java.util.ArrayList;

public class BeheerBoerderijenController {

    @FXML
    private Button btnDelete;
    @FXML
    private Button btnAdd;
    @FXML
    private Button btnModify;
    @FXML
    private Button btnClose;
    @FXML
    private TableView tblBoerderijen;

    public void initialize() {
        initTable();
        btnAdd.setOnAction(e -> addNewRow());
        btnModify.setOnAction(e -> {
            verifyOneRowSelected();
            modifyCurrentRow();
        });
        btnDelete.setOnAction(e -> {
            verifyOneRowSelected();
            deleteCurrentRow();
        });
        
        btnClose.setOnAction(e -> {
            var stage = (Stage) btnClose.getScene().getWindow();
            stage.close();
        });
    }

    private void initTable() {
        tblBoerderijen.getSelectionModel().setSelectionMode(SelectionMode.SINGLE);
        tblBoerderijen.getColumns().clear();

        // TODO verwijderen en "echte data" toevoegen!
        int colIndex = 0;
        for(var colName : new String[]{"Naam", "Voornaam", "Oppervlakte", "Aantal varkens"}) {
            TableColumn<ObservableList<String>, String> col = new TableColumn<>(colName);
            final int finalColIndex = colIndex;
            col.setCellValueFactory(f -> new ReadOnlyObjectWrapper<>(f.getValue().get(finalColIndex)));
            tblBoerderijen.getColumns().add(col);
            colIndex++;
        }

        for(int i = 0; i < 10; i++) {
            tblBoerderijen.getItems().add(FXCollections.observableArrayList("Boer " + i, "Jozef V" + i, i*10 + "", i * 33 + ""));
        }
    }

    private void addNewRow() {
    }

    private void deleteCurrentRow() {
    }

    private void modifyCurrentRow() {
    }

    public void showAlert(String title, String content) {
        var alert = new Alert(Alert.AlertType.WARNING);
        alert.setTitle(title);
        alert.setHeaderText(title);
        alert.setContentText(content);
        alert.showAndWait();
    }

    private void verifyOneRowSelected() {
        if(tblBoerderijen.getSelectionModel().getSelectedCells().size() == 0) {
            showAlert("Hela!", "Eerst een boer selecteren hé.");
        }
    }
}

File: examples/kotlin/project-template/src/main/kotlin/be/kuleuven/csa/controller/BeheerBoerderijenController.kt
package be.kuleuven.csa.controller

import javafx.collections.FXCollections

import javafx.beans.property.ReadOnlyObjectWrapper

import javafx.collections.ObservableList

import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.stage.Stage


class BeheerBoerderijenController {
    @FXML
    private lateinit var btnDelete: Button

    @FXML
    private lateinit var  btnAdd: Button

    @FXML
    private lateinit var  btnModify: Button

    @FXML
    private lateinit var  btnClose: Button

    @FXML
    private lateinit var  tblBoerderijen: TableView<ObservableList<String>>

    fun initialize() {
        initTable()
        btnAdd.setOnAction { e -> addNewRow() }
        btnModify.setOnAction { e ->
            verifyOneRowSelected()
            modifyCurrentRow()
        }
        btnDelete.setOnAction { e ->
            verifyOneRowSelected()
            deleteCurrentRow()
        }
        btnClose.setOnAction { e ->
            val stage = btnClose.getScene().getWindow() as Stage
            stage.close()
        }
    }

    private fun initTable() {
        tblBoerderijen.getSelectionModel().setSelectionMode(SelectionMode.SINGLE)
        tblBoerderijen.getColumns().clear()

        // TODO verwijderen en "echte data" toevoegen!
        var colIndex = 0
        for (colName in arrayOf("Naam", "Voornaam", "Oppervlakte", "Aantal varkens")) {
            val col: TableColumn<ObservableList<String>, String> = TableColumn(colName)
            val finalColIndex = colIndex
            col.setCellValueFactory { f -> ReadOnlyObjectWrapper(f.getValue().get(finalColIndex)) }
            tblBoerderijen.getColumns().add(col)
            colIndex++
        }
        for (i in 0..9) {
            tblBoerderijen.getItems().add(
                FXCollections.observableArrayList(
                    "Boer $i",
                    "Jozef V$i",
                    (i * 10).toString(),
                    (i * 33).toString()
                )
            )
        }
    }

    private fun addNewRow() {}

    private fun deleteCurrentRow() {}

    private fun modifyCurrentRow() {}

    fun showAlert(title: String?, content: String?) {
        val alert = Alert(Alert.AlertType.WARNING)
        alert.title = title
        alert.headerText = title
        alert.contentText = content
        alert.showAndWait()
    }

    private fun verifyOneRowSelected() {
        if (tblBoerderijen.getSelectionModel().getSelectedCells().size === 0) {
            showAlert("Hela!", "Eerst een boer selecteren hé.")
        }
    }
}